File: python/word_squares.py
class TrieNode:
    def __init__(self):
        self.prefix = []
        self.child ={}


class Solution:
    """
    @param: words: a set of words without duplicates
    @return: all word squares
    """
    def __init__(self):
        self.root = TrieNode()
        
    def buildTrie(self, words):
        for i in range(len(words)):
            tmp = self.root
            for j in range(len(words[i])) :
                if words[i][j] not in tmp.child:
                    tmp.child[words[i][j]] = TrieNode()
                tmp.prefix.append(i)
                tmp = tmp.child[words[i][j]]
    
    
    def wordSquares(self, words):
        # write your code here
        res = []
        if not words or not words[0]:
            return res
            
        self.buildTrie(words)
        
        out = []
        for i in range(len(words)):
            out.append(words[i])
            self.getPrefix(words, out, res)
            out.pop()
            
        return res
        
    
    def getPrefix(self, words, out, res):
        if len(out) == len(words[0]):
         
            res.append(list(out))
            return
       
        n = len(out)
        # for i in len(out):
        #     target += out[i][n]
            
        target = ''.join([out[i][n] for i in range(n)])
        
        tmp = self.root
        for j in range(len(target)) :
            if target[j] not in tmp.child :
                return
            tmp = tmp.child[target[j]]
        
        for k in range(len(tmp.prefix)):
            out.append(words[tmp.prefix[k]])
            self.getPrefix(words, out, res)
            out.pop()
            
        
test = Solution()
words = ["area", "lead", "wall", "lady", "ball"]
res = test.wordSquares(words)
print(res)

File: cpp/favourite/WordSquares.cpp
struct TrieNode{

    TrieNode* child[26];
    vector<int> prefix;
    TrieNode(){
      
        for (int i=0; i<26; i++) {
            child[i] = nullptr;
        }
    }
};



class Solution {
    TrieNode* root;
public:
    /*
     * @param words: a set of words without duplicates
     * @return: all word squares
     */
    
    Solution(){
        root = new TrieNode();
    }
    
    void buildTrie(vector<string>& words) {
        
        for(int j=0; j<words.size(); j++) {
           TrieNode* tmp = root;
           for(int i =0; i<words[j].size(); i++) {
               
               if(!tmp->child[words[j][i]-'a']){
                   tmp->child[words[j][i]-'a'] = new TrieNode();
               }
               tmp->prefix.push_back(j);
               tmp = tmp->child[words[j][i]-'a'];
           }
          
          
        }
    }
    
    vector<vector<string>> wordSquares(vector<string> &words) {
        // write your code here
        vector<vector<string>> res;
        int m = words.size();
        if(!m) return res;
        buildTrie(words);
        
        vector<string> out;
        for(int i=0; i<m; i++) {
            out.push_back(words[i]);
            getSquare(out, words, res);
            out.pop_back();
        }
        
        return res;
        
    }
    
    void getSquare(vector<string>& out, vector<string>& words, vector<vector<string>>& res) {
        if(out.size() == words[0].size()) {
            res.push_back(out);
            return;
        }
        
        int sz = out.size();
        string target;
        for (int i=0; i<sz; i++) {
            target += out[i][sz];
        }
        
        auto tmp = root;
        for (int i=0; i<target.size(); i++) {
            if(!tmp->child[target[i]-'a']) return;
            tmp = tmp->child[target[i]-'a'];
        }
        
        vector<string>next;
        for (int i=0; i<tmp->prefix.size(); i++) {
            next.push_back(words[tmp->prefix[i]]);
        }
        
        for (auto nt : next) {
            out.push_back(nt);
            getSquare(out, words, res);
            out.pop_back();
        }
   
        
    }
 
    
    
    
};